private/library.rkt

```diff
@@ -0,0 +1,297 @@
+;; to maintain compatibility.
+;; Therefore:
+;;   - For display to users, we preserve the distinctions among directories
+;;     using path->relative-string/library, which includes package information.
+;;   - For persistent storage, we represent an collection-based exclusion as a
+;;     normalized-lib-module-path?, which will continue to apply regardless of
+;;     what package (or even direct link) supplies the collection.
+;;   - The set of collection-based script directories is already stored as part
+;;     of the Racket installation (in the info.rkt files and caches).
+;;     We do not store it again with our saved state.
+
+;; the library data we save (shared across Racket versions)
+(serializable-struct
+ ;; can evolve in the future using serializable-struct/versions
+ library-data (table collection-based-exclusions)
+ #:guard (struct-guard/c
+          (and/c (hash/c #:immutable #t
+                         #:flat? #t
+                         path?
+                         (set/c path-element? #:cmp 'equal-always))
+                 hash-equal-always?)
+          (set/c (and/c normalized-lib-module-path?
+                        (list/c 'lib immutable-string?))
+                 #:cmp 'equal-always))
+ #:transparent)
+(define (make-library-data)
+  (library-data (hashalw user-script-dir (setalw))
+                (setalw)))
+
+;; a wrapper with installation info and some caches
+;; this is NOT thread-safe, due to hash mutation
+(struct library (lib collects-dirs setup-cache mp-cache pretty-cache)
+  #:transparent)
+(define (library-data->library lib)
+  (library lib
+           (find-collection-based-script-directories)
+           (make-hash)
+           (make-hash)
+           (make-hash)))
+
+(define find-collection-based-script-directories
+  (let ([absent (gensym)])
+    (define (find-collection-based-script-directories)
+      (for*/setalw ([dir (find-relevant-directories '(quickscript-directory))]
+                    [info (in-value (get-info/full dir))]
+                    #:when info
+                    [v (in-value (info 'quickscript-directory (λ () absent)))]
+                    #:when (cond
+                             [(eq? #t v)]
+                             [else
+                              (unless (eq? absent v)
+                                (log-quickscript-error
+                                 "~a\n  expected: ~e\n  given: ~e\n  directory: ~e"
+                                 "bad value for quickscript-directory in info file"
+                                 #t
+                                 v
+                                 dir))
+                              #f]))
+        dir))
+    find-collection-based-script-directories))
+
+;; library-data is stored using the framework/preferences system,
+;; which provides help for future changes without breaking compatibility
+(define pref-key 'plt:quickscript:library)
+(preferences:set-default pref-key (make-library-data) library-data?)
```

Comment:
Should we initialize the preference with an existing `library.rktd` file if one exists? One reason *not* to do so is that we would carry forward the duplication of existing directories as in https://github.com/Metaxal/quickscript-extra/issues/27, which this change would otherwise fix.

Reply:
The more I think about it, the more I think the proper and simpler fix looks like this:

* For installed collections, scripts are added automatically and only exclusions (either directories or single scripts) are stored. This avoids both the duplication issue and the register-on-setup problem.
* For paths added manually by the user, store the included paths and the excluded scripts (as is currently the case)

This can be done with only little change the library system, independently of the preferences system.

Reply:
I think this comment is what this PR currently does, except that there is no mechanism for excluding whole directories. That's a feature we could consider, but it raises more questions, and it seemed to me that either uninstalling the unwanted package or excluding all of the scripts individually would be good enough for most cases.

It's true that changing from `library.rktd` to the preferences system isn't directly related to the rest of this change. I don't think we can just continue using `library.rktd`, though, because it needs to be able to be understood by old versions of Racket, and we want to store at least some information that old versions can't handle. We could start using `library- v2.rktd` or something, but this compatibility problem is exactly the same as applies to the rest of the preferences system, which is why I suggested adopting that already-existing solution.

What I was trying to ask here is whether, if new-style data hasn't been saved (regardless of how we decide to save it), we should try to import an old `library.rktd` file if one exists. The more I think about it, the more strongly I think not. There seem to be three categories of entries that might be present in `library.rktd`:

1. `user-script-dir`, which we will add by default anyway. We could import any exclusions from `user-script-dir` if that seemed worthwhile.
2. Collection directories from an old Racket version, perhaps including duplicates due to https://github.com/Metaxal/quickscript-extra/issues/27. Importing these would be a bug! As long as the corresponding package is installed, the new mechanism will take care of these. Unfortunately I don't see any good way to identify which entries these are, which means I don't know how we could import only the exclusions, if any.
3. Non-collection directories explicitly added by the user (i.e. *not* `user-script-dir`) are what would be nicest to be able to import, but we have no way of distinguishing these from the collection-based directories that should *not* be imported.

I think this means we have the choice of two kinds of less-than-ideal behavior: either we import old collection/package directories that we shouldn't; or we omit explicitly added directory entries that we should. I think it's better to avoid the wrong imports. I expect that more people have gotten entries in `library.rktd` from this package in 8.11 or `quickscript-extra` at any version than have explicitly added directories other than `user-scripts-dir`. Furthermore, anyone who has explicitly added directories has used the Quickscript library manager to do so, and can be asked to do so again, whereas someone who only got scripts from packages may never have used the Quickscript library manager.

Reply:
This is a one-off, so as long as this breaking change is well advertized I'm mostly okay with letting the user reconfigure their library.

If that's not too much effort we could still consider item 1. though, which should be the most common one.
It would be nice to treat exclusions from `quickscript-extra` and `quickscript-competition-2020` specifically, if that's not too much trouble, because of shadow scripts — otherwise this would lead to duplicate entries in the menu. But again, I don't feel too strongly on this, and it will depend on the required effort.

Reply:
I've pushed further commits to import old exclusions for `user-script-dir` by default.

I gave some thought to heuristics to find other exclusions we'd want to import, but, the more I think about it, the more strongly I think the right thing to do is to provide documentation for the user. Even just considering exclusions from `quickscript-extra` due to shadow scripts, even assuming we could reliably recognize it and ignoring https://github.com/Metaxal/quickscript-extra/issues/27, importing the exclusion wouldn't be enough to fully solve the problem: the user would also need to edit the shadow script itself, because the old script would use something like:

```racket
(require (file "/absolute/path/to/THE/OLD/VERSION/OF/quickscript-extra/scripts/tweet.rkt"))
```

(This PR fixes that going forward by using collection-based module paths when shadowing collection-based scripts.)

Reply:
Thanks for all this! 

It doesn't seem like there is a seamless solution for the user unfortunately, as they will be left with old shadow scripts to delete or adapt and disable the corresponding original scripts, or new shadow scripts to rewrite.

But the previous state with duplicate registered directories was bad either way. So your approach is the  best I can think of for now. I'm just worried that some users will miss the announcement about what to do. Maybe we should include a section in the docs about pre-to-post Racket 8.13.